Issue #1087 redis fix

matteius opened the pull request:
This is the start of a PR to address issues introduced with:  https://github.com/celery/kombu/pull/1068/files
Reported as:
https://github.com/celery/kombu/issues/1087
https://github.com/celery/kombu/issues/1081
Possibly others ...

This does not revert the entire change from @auvipy and instead assumes that Redis implementation.  I would be curious of anyone with bug reports tried if that resolves their issue.  I also would be curious to hear from auivpy more about this change.  I see the intent was some kind of optimization O(1) vs O(n) but was this ever actually a performance issue?   Do we know for sure that the _lookup_direct method being called from base now is working well?  I would assume so but also we haven't upgraded up to this version that was affected yet.

matteius commented:
I'll take a look again tomorrow -- time for rest.

auvipy commented:
this is the main issue https://github.com/celery/kombu/issues/994 it would e great if you could check this first.

thedrow commented:
@matteius This was a potential performance issue.
However, it didn't solve #994.

I still think we need to figure out why it breaks our Redis broker implementation.

matteius commented:
I am going to do more analysis soon, but my hunch is that the method signature changed to not contain a default -- but also its confusing code because we are cross pollinating discussion about Redis data structures but the code changes seemed to change even how we communicate with the sockets beyond just how we use the data structures underneath Redis.  The sun is shining and I've got pressure washing to do, so I'll check back later.

matteius commented:
Okay -- waiting for the appveyor build to catch up but I think we have something here, I am not entirely sure what test case I would add to demonstrate the changes here, because partly I am trying to fix the Redis bug introduced in 4.6.4.  I did not just want to back out @auvipy changes because I think he made some good progress on self documenting the code with variable names and adding the O(1) vs O(n) enhancement.  Also I did not want to keep the suggestion yet for a private _lookup_direct method series because prior we had just the one base _lookup method, and then ended up 3 private methods in 4.6.4 related to _lookup -->  I think having two right now we can try to test where this one is a redis specific implementation, will be useful.  I kept parity between the try/except error handling and log warnings between the two methods.

thedrow commented on kombu/transport/redis.py:
```diff
@@ -829,23 +831,45 @@ def get_table(self, exchange):
+            with self.conn_or_acquire() as client:
+                if client.sismember(redis_key, queue_bind):
+                    result = [queue]
+                if not client.scard(redis_key):
```
Why is this below `sismember`?
Won't `sismember` return false anyway if `redis_key`'s cardinality is zero?

matteius replied:
Yup good catch -- I mean't for it to go first.

thedrow commented on kombu/transport/redis.py:
```diff
@@ -829,23 +831,45 @@ def get_table(self, exchange):
+                if client.sismember(redis_key, queue_bind):
+                    result = [queue]
+                if not client.scard(redis_key):
+                    raise InconsistencyError(NO_ROUTE_ERROR.format(
```
Are we sure we want to raise an exception instead of returning `[]` as before?
Why change this?

See https://github.com/celery/kombu/blob/v4.3.0/kombu/transport/virtual/base.py#L718 for example.

matteius replied:
Yes, it is subtle but that method you referenced calls lookup which has as one of its arguments self.get_table(exchange) -- which will raise this InconsistencyError if there are no members in the set matching that key.   This is why I generally prefer that code not be so deeply nested -- I'd prefer that the call in base define the variable name that is get_table and pass that in, because it becomes self documenting code -- its not clear initially to the reader without much analysis that get_table is going to return all members in the set, nor is it clear that the call to lookup might raise an InconsistencyError but really anything that might raise its own error should probably be on its own line instead of tightly nested in a function call.

matteius replied:
https://github.com/celery/kombu/blob/master/kombu/transport/redis.py#L829

matteius replied:
What I am mostly unclear on is if the patrern is actually always going to be an empty string '' for redis?

matteius commented:
I will soon -- I need to get some rest, but I started adding a couple redis test cases, I just haven't perfected them yet.  Back soon.

thedrow commented on t/unit/transport/test_redis.py:
```diff
@@ -33,151 +35,104 @@ def poll(self, timeout):
 
-    def sadd(self, key, member, *args):
-        self.sets[key].add(member)
+    # def send_command(self, cmd, *args):
```
If its commented out just remove it.

matteius commented:
Ok this has been a lot of work, but I think it will end up being worth it.  I had check point my progress or I wasn't going to be able to think about anything else ever again.  As a result, I commented out two test Redis test cases that are still failing for me -- test_publish__get and test_publish__consume.  

The problem I ran into here, and why I probably have going on 30 hours on this single effort, is that in refactoring one things and trying to add some tests, I found that the test mocks were insufficient.  I did not like much what we had for a FakeRedisClient and it was mostly duplicated and partial from fakeredis.  Then I went on to learn that fakeredis is partial from what redis actually implements because it doesn't do much at all for pipeline execution although everything else works pretty well from that library.  I had to extend from some of those mocks in kombu as well in order to support the Channel and Transport concepts--as such I've added a FakeRedisKombuChannel and a FakeRedisKombuTransport.  In addition to providing the following fakeredis extensions:  FakeRedisSocket, FakeRedisConnection, FakeRedisConnectionPool, FakeRedisClient, my own FakePipeline concept to support the other fakeredis stuff, and renamed the old mock Pipeline class to FakePipelineLite. 

I need to revisit the two commented out tests as I think they are conceptually important but I need to study them more.  At least getting a fully passing CI build lets me regain confidence in my efforts here both with myself and with the core developers, as well as clean up my linting errors and made myself delete the commented out blocks and the 1000+ pprint statements I added because I haven't remembered in a while why I know how to use an IDE/debugger for next time.

Anyway, this is ready for a more full fledged review and PR feedbacks now.

thedrow commented:
Can you please check if you resolved #994?

matteius commented:
@thedrow  I really don't have time to run a code profile on it anytime soon, the next bit of time I do have will be to revisit those two unit tests and the decrease in coverage of kombu/transport/redis.py | 82.18% <84.44%> (-8.27%)

Perhaps @stevanmilic can re-run a trace using this branch and also to check if this code works in their code path?  Might be easier for you since you ran the original trace on this code.  To be honest I don't actually use the kombu redis backend, but I have a bunch of Redis experience and its a pet interest of mine, so I wanted to help out here in that space as well.

The concepts that @auvipy implemented in the change to use sismember is here in my PR, but it holds more true to the original _lookup method and it was that prior change that broke causing the issue report for this PR.  I think this mends both worlds together in a working well, but could use help testing a real integration outside of faked out unit tests.  Thanks!

matteius commented:
@auvipy Yeah you would not be able to make sense of it if you go commit by commit now -- I can get this branch down to a single commit though.   The changes are not very much so large to review--most of the changes are to the tests and once I correct the code coverage issue this build will be fully passing your CI.  Changes that are included not related to test files are:

- Updated _list to be more like it was before but using the proposed redis commands optimization to be O(1) -- this removes the broken _lookup_direct method and restores a consistent _lookup api.
- Removed any code paths for redis.VERSION[0] < 3 since we assert and require at least 3 already.
- Renamed certain method and variables their usages (easily using pycharm refactor) to have readable easy to understand method names that don't require a pHD in English acronyms to understand, method renamed _q_for_pri --> _queue_for_priority.
- Requires fakeredis now for tests

and thats it, production facing code changes here are very minimal and fixes a bug.

matteius commented:
@auvipy Alright I recreated my branch with minimal commit history of just the diff of current changes, and then I added some more unit test coverage to get the coverage check to pass.  Right now all 4 checks pass.  My only TODO is really to revisit those two unit test cases I have commented out because they don't work right my current mocks -- but I plan to do something more with those.  Otherwise though I can't resolve your change you requested because its tied to a commit that no longer exists as I force recreated my branch in order to satisfy your requested change.

auvipy commented on t/unit/transport/test_redis.py:
```diff
@@ -835,88 +953,86 @@ def test_rediss_connection(self):
+    #     assert self.queue(channel).get() is None
+    #     assert self.queue(channel).get() is None
+
+    # @skip("This test is currently broken")
```
disabling these two tests should be removed? or kept for future improvement? don't they affect any current code?

auvipy replied:
OK

matteius replied:
I've moved these tests down to a new test class that is based on mocks like they originally used.  Now the tests pass -- will see how the latest build goes, maybe a flake8 error to fix, maybe not.

matteius commented:
Hey guys -- @thedrow and @auvipy -- I've finished all my TODOs and restored those two commented out test cases as their own new TestCase class that utilizes Lite mocks like the ones that were in the test_redis.py file prior.  All 4 CI checks have passed successful--awaiting your final approvals.

auvipy commented on t/unit/transport/test_redis.py:
```diff
@@ -229,42 +444,97 @@ def _get_response_error(self):
+
 @skip.unless_module('redis')
-class test_Channel:
+class TestRedisChannel(unittest.TestCase):
```
can't we use pytest alternative here?

matteius replied:
I mean probably, although I didn't understand why kombu includes its own version that it already uses everywhere, and I didn't want to introduce too much scope creep for this already overly large set of changes/code diff.

matteius replied:
@auvipy I am proposing it is a good suggestion that we open as a new ticket, after the next kombu release (or you can open the ticket now).  However I want to suggest something that I think could be awesome for this project and the other Celery dependencies, but lets start with Kombu.  Following the next release of Kombu I would like us to find like 2 determine and schedule two time slots a week for up to 1-hour each that are recurring optional invites and when available we join and groom the Issues report.  By treating that backlog as an Agile scrum backlog of sorts we can clean it up as a group, I mean if myself or you or someone else knows of issues that were fixed and we can relate that back to open issues in a group session and determine which tickets to close and update we can get the list much much smaller than in a shorter amount of time than working entirely disparately.   

Also, for me I became an Agile CSM -- Certified SCRUM Master in October of 2015 and have been serving as a Developer and CSM for my team at work since that time.  They recently changed requirements for renewal of the certification to do additional continuing education units, and I am hoping to petition them that opens source work such as this should be able to count for some of those hours when paying the renewal fees.  Let me know what you think, we can get some kind of WexEx link that we can join and crush this backlog in prep for Celery 5!

matteius commented:
@auvipy  Are you able to resolve that old change requested that links off to a commit that no longer exists?  Is that how we get the code coverage bot to post an updated report because the last comment from the bot is definitely way stale at this point.
